les/lespay/server/prioritypool.go

```diff
@@ -0,0 +1,502 @@
+// by externally resetting both flags. ActiveFlag should not be set externally.
+//
+// The highest priority nodes in "inactive" state are moved to "active" state as soon as
+// the minimum capacity can be granted for them. The capacity of lower priority active
+// nodes is reduced or they are demoted to "inactive" state if their priority is
+// insufficient even at minimal capacity.
+type PriorityPool struct {
+	PriorityPoolSetup
+	ns                     *nodestate.NodeStateMachine
+	clock                  mclock.Clock
+	lock                   sync.Mutex
+	activeQueue            *prque.LazyQueue
+	inactiveQueue          *prque.Prque
+	changed                []*ppNodeInfo
+	activeCount, activeCap uint64
+	maxCount, maxCap       uint64
+	minCap                 uint64
+	activeBias             time.Duration
+	capacityStepDiv        uint64
+}
+
+// nodePriority interface provides current and estimated future priorities on demand
+type nodePriority interface {
+	// Priority should return the current priority of the node (higher is better)
+	Priority(now mclock.AbsTime, cap uint64) int64
+	// EstMinPriority should return a lower estimate for the minimum of the node priority
+	// value starting from the current moment until the given time. If the priority goes
+	// under the returned estimate before the specified moment then it is the caller's
+	// responsibility to signal with updateFlag.
+	EstMinPriority(until mclock.AbsTime, cap uint64, update bool) int64
+}
+
+// ppNodeInfo is the internal node descriptor of PriorityPool
+type ppNodeInfo struct {
+	nodePriority               nodePriority
+	node                       *enode.Node
+	connected                  bool
+	capacity, origCap          uint64
+	bias                       time.Duration
+	forced, changed            bool
+	activeIndex, inactiveIndex int
+}
+
+// NewPriorityPool creates a new PriorityPool
+func NewPriorityPool(ns *nodestate.NodeStateMachine, setup PriorityPoolSetup, clock mclock.Clock, minCap uint64, activeBias time.Duration, capacityStepDiv uint64) *PriorityPool {
+	pp := &PriorityPool{
+		ns:                ns,
+		PriorityPoolSetup: setup,
+		clock:             clock,
+		activeQueue:       prque.NewLazyQueue(activeSetIndex, activePriority, activeMaxPriority, clock, lazyQueueRefresh),
+		inactiveQueue:     prque.New(inactiveSetIndex),
+		minCap:            minCap,
+		activeBias:        activeBias,
+		capacityStepDiv:   capacityStepDiv,
+	}
+
+	ns.SubscribeField(pp.priorityField, func(node *enode.Node, state nodestate.Flags, oldValue, newValue interface{}) {
+		if newValue != nil {
+			c := &ppNodeInfo{
+				node:          node,
+				nodePriority:  newValue.(nodePriority),
+				activeIndex:   -1,
+				inactiveIndex: -1,
+			}
+			ns.SetFieldSub(node, pp.ppNodeInfoField, c)
+		} else {
+			ns.SetStateSub(node, nodestate.Flags{}, pp.ActiveFlag.Or(pp.InactiveFlag), 0)
+			ns.SetFieldSub(node, pp.ppNodeInfoField, nil)
```

Comment:
I would really like that we move the `disconnection` logic here. 

So that it means this callback is the **only** place which can initialize the `ppNodeInfoField` or set it to nil.

Reply:
done

Reply:
I made a slight change to NSM because the node dropping logic kind of screwed this concept:
https://github.com/ethereum/go-ethereum/pull/21436#issuecomment-686146352
Now I think both NSM and this logic are simpler.

Reply:
I like this change(if any flag or field is available, then don't drop the node).